acceptance_test.go

```diff
@@ -0,0 +1,186 @@
+				NewSource:        source.NewSource,
+				NewDestination:   destination.NewDestination,
+			},
+			SourceConfig:      sourceConf,
+			DestinationConfig: destConf,
+			GoleakOptions:     []goleak.Option{goleak.IgnoreCurrent()},
+			Skip: []string{
+				// It calls Write & WriteAsync function directly, with non structure data generated using random strings
+				// and there is no option to substitute existing generateRecords function
+				"TestDestination_Write_Success",
+			},
+			AfterTest: func(t *testing.T) {
+				mr.FlushAll()
+			},
+		},
+		},
+		rand: rand.New(rand.NewSource(time.Now().UnixNano())), // nolint: gosec // only used for testing purpose
+	})
+}
+
+type AcceptanceTestDriver struct {
+	rand *rand.Rand
+	sdk.ConfigurableAcceptanceTestDriver
+}
+
+func (d AcceptanceTestDriver) WriteToSource(t *testing.T, records []sdk.Record) []sdk.Record {
+	if d.Connector().NewDestination == nil {
+		t.Fatal("connector is missing the field NewDestination, either implement the destination or overwrite the driver method Write")
+	}
+
+	is := is.New(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// writing something to the destination should result in the same record
+	// being produced by the source
+	dest := d.Connector().NewDestination()
+	// write to source and not the destination
+	destConfig := d.SourceConfig(t)
+	err := dest.Configure(ctx, destConfig)
+	is.NoErr(err)
+
+	err = dest.Open(ctx)
+	is.NoErr(err)
+	records = d.generateRecords(len(records))
+	// try to write using WriteAsync and fallback to Write if it's not supported
+	err = d.write(ctx, dest, records)
+	is.NoErr(err)
+
+	cancel() // cancel context to simulate stop
+	err = dest.Teardown(context.Background())
+	is.NoErr(err)
+
+	return records
+}
+
+// write writes records to destination using Destination.Write.
+func (d AcceptanceTestDriver) write(ctx context.Context, dest sdk.Destination, records []sdk.Record) error {
+	for _, r := range records {
+		err := dest.Write(ctx, r)
+		if err != nil {
+			return err
+		}
+	}
+
+	// flush to make sure the records get written to the destination, but allow
+	// it to be unimplemented
+	err := dest.Flush(ctx)
+	if err != nil && !errors.Is(err, sdk.ErrUnimplemented) {
+		return err
+	}
+
+	// records were successfully written
+	return nil
+}
+
+func (d AcceptanceTestDriver) generateRecords(count int) []sdk.Record {
+	records := make([]sdk.Record, count)
+	for i := range records {
+		records[i] = d.generateRecord()
+	}
+	return records
+}
+
+func (d AcceptanceTestDriver) generateRecord() sdk.Record {
+	return sdk.Record{
+		Metadata: map[string]string{
+			"key": sourceKey,
+		},
+		CreatedAt: time.Time{},
+		Key:       sdk.RawData(sourceKey),
+		Payload:   sdk.RawData(fmt.Sprintf(`{"%s":"%s"}`, d.randString(32), d.randString(32))),
+	}
+}
+
+// randString generates a random string of length n.
+// (source: https://stackoverflow.com/a/31832326)
+func (d AcceptanceTestDriver) randString(n int) string {
```

Comment:
why did you generate your own records?

Reply:
Because stream destination expects the data to be valid JSON format, whereas predefined generator only generates random string. Also, we needed to specify the data in Key & Position to match what will be expected from the Source connector.